### PR TITLE
[WIP] Fixes #25756 - Adds extend-ability to react components

### PR DESCRIPTION
**Pull request by amirfefer**
### Work In Progress
This PR is written according to [this](https://community.theforeman.org/t/rfc-expandable-component-architecture/12083/12) design
ezgif com-video-to-gif

In this demonstration I've added `<Fill>` component in katello subscription details page, and a `<Slot>` in the upper left bar:
```js
        <Fill id="test">
          <Nav navbar pullLeft className="navbar-iconic">
            <Button style={{ marginTop: '15px' }} bsStyle="primary"> Extenable Button !</Button>
          </Nav>
        </Fill>
```

### CODE REVIEW

**Comment by amirfefer** [comment]
**UPDATE:** a slot can handle multiple Fills components, the rendering order will be by weights:
screenshot from 2018-12-26 18-31-17

File: webpack/assets/javascripts/react_app/components/common/Fill/ExtendableRegistery.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { orderBy } from 'lodash';
+
+const components = {};
+
+export const registerComponent = (SlotId, key, component, weight) => {
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/Fill/ExtendableRegistery.js]: would register/unregister (or add/remove etc) are enough? I think the component should already part of the context, e.g. when you are using it:
```js
import { add, remove } from './ExtendableRegistery';
```

File: webpack/assets/javascripts/react_app/components/common/Fill/ExtendableRegistery.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  if (components[SlotId]) {
+    components[SlotId].forEach((item, index) => {
+      if (item.key === key) {
+        components[SlotId].splice(index, 1);
+      }
+    });
+  }
+};
+
+export const getComponent = id =>
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/Fill/ExtendableRegistery.js]: maybe worth comparing with the ruby implementation for menus e.g. at https://github.com/theforeman/foreman/blob/develop/app/registries/menu/manager.rb

File: webpack/assets/javascripts/react_app/components/common/Fill/Fill.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+class Fill extends React.Component {
+  componentDidMount() {
+    const { children, registerFillComponent, id, weight, fillId } = this.props;
+    debugger;
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/Fill/Fill.js]: oops

File: webpack/assets/javascripts/react_app/components/common/Fill/Fill.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    registerFillComponent(id, fillId, children, weight);
+  }
+  componentWillUnmount() {
+    const { id, removeRegisteredComponent, fillId } = this.props;
+    removeRegisteredComponent(id, fillId);
+  }
+  render() {
+    return <div />;
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/Fill/Fill.js]: TODO?

**Review by ohadlevy — COMMENTED** [review]
Looks like a good start! I've added a few comments inline, plus there is an empty reducer file.

**Comment by glekner** [comment]
This is amazing!

Would be great to have an option to give props to `<Fill />` instead of JSX.
I guess this means every Slot will need to have a default Fill in it that can receive props for it to be customizable.

**Comment by amirfefer** [comment]
Great idea @glekner ! 
As you suggested, a `Fill` can contain JSX or plain object as props, with this logic plugins may totally override a slot or just modify the props.

#### Props way
```js
// plugin
<Fill id='slot-id' fillId='some-id' overrideProps={{ text: 'this text given by a prop' }} /> 
```

```js
// foreman core
const Wrapper = ({ text }) => <div>{text}</div>; 
<Slot id='slot-id'>
  <Wrapper text='some default text' />
</Slot>
```
#### JSX way
```js
// plugin
<Fill id='slot-id' fillId='some-id'> 
  <div> some text </ div>
</ Fill>
```

```js
// foreman core
<Slot id='slot-id'>
 default children // can be empty
</Slot>
```

File: webpack/assets/javascripts/react_app/components/common/Fill/ExtendableRegistery.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  }
+  if (components[SlotId] === undefined) {
+    components[SlotId] = [];
+  }
+  components[SlotId].push({ key, component, weight });
+};
+
+export const remove = (SlotId, key) => {
+  if (components[SlotId]) {
+    components[SlotId].forEach((item, index) => {
```
Comment (xprazak2) [inline webpack/assets/javascripts/react_app/components/common/Fill/ExtendableRegistery.js]: nit: have you considered using [filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter)?

File: webpack/assets/javascripts/react_app/components/common/Fill/Fill.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+class Fill extends React.Component {
+  componentDidMount() {
+    const { children, registerFillComponent, id, weight, fillId } = this.props;
+
+    registerFillComponent(id, fillId, children, weight);
+  }
+  componentWillUnmount() {
+    const { id, removeRegisteredComponent, fillId } = this.props;
+
+    removeRegisteredComponent(id, fillId);
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/Fill/Fill.js]: nitpick: `unregisterFillComponent`?

File: webpack/assets/javascripts/react_app/components/common/Slot/Slot.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  getComponentByWeight,
+  getMaxComponent,
+} from '../Fill/ExtendableRegistery';
+
+const Slot = ({ shouldUpdate, id, multi, children, ...props }) => {
+  const addProps = object => {
+    if (isValidElement(object)) {
+      return cloneElement(object, props);
+    }
+    return cloneElement(children, object);
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/Slot/Slot.js]: shouldn't you keep the old props? `return cloneElement(children, { ...props, ...object });`

**Review by sharvit — COMMENTED** [review]
Awesome work @amirfefer :1st_place_medal:

File: webpack/assets/javascripts/react_app/components/common/Fill/ExtendableRegistery.js (hunk 5 above)
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/Fill/ExtendableRegistery.js]: filter creates a new array, while using splice the removing process is in place.
we can use filter and splice, having something like:
```js
obj.splice(0, obj.length, ...obj.filter(item => item.key !== key))
```

File: webpack/assets/javascripts/react_app/components/common/Fill/Fill.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+class Fill extends React.Component {
```
Comment (xprazak2) [inline webpack/assets/javascripts/react_app/components/common/Fill/Fill.js]: should be added to component registry, same for Slot

File: webpack/assets/javascripts/react_app/components/common/Fill/index.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+});
+
+// map action dispatchers to props
+const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);
+
+// export reducers
+export const reducers = { extends: reducer };
+
+// export connected component
+export default connect(
```
Comment (xprazak2) [inline webpack/assets/javascripts/react_app/components/common/Fill/index.js:21]: should be added into component registry, same for Slot

File: webpack/assets/javascripts/react_app/components/common/Fill/FillActions.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
+import { REGISTER_FILL, REMOVE_FILLED_COMPONENT } from './FillConstants';
+import { add, remove } from './ExtendableRegistery';
```
Comment (xprazak2) [inline webpack/assets/javascripts/react_app/components/common/Fill/FillActions.js:2]: [Google](https://www.google.com/search?client=firefox-b&q=Dictionary#dobs=registry) knows only the version without 'e', could you rename please?

File: webpack/assets/javascripts/react_app/components/common/Fill/FillReducer.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+  switch (action.type) {
+    case REGISTER_FILL:
+      return state.setIn(
+        ['RegisteredComponent', payload.slotId, payload.fillId],
+        true
+      );
+
+    case REMOVE_FILLED_COMPONENT:
+      return state.setIn(
+        ['RegisteredComponent', payload.slotId, payload.fillId],
```
Comment (xprazak2) [inline webpack/assets/javascripts/react_app/components/common/Fill/FillReducer.js:21]: Same as [comment](https://github.com/Katello/katello/pull/8039#discussion_r268562069) on the other end.

File: app/views/about/index.html.erb [hunk 12]
```diff
@@ -141,4 +141,5 @@
     </div>
   </div>
 </div>
-</div>
+
+<%= slot('about-footer-slot', true) %>
```
Comment (xprazak2) [inline app/views/about/index.html.erb:145]: Needs to be inside `div#about` tag, so it does not show all the way down:

**Comment by waldenraines** [comment]
@amirfefer can you rebase this so I can test it out please?

**Comment by amirfefer** [comment]
I'm closing this PR because I opened another one
 https://github.com/theforeman/foreman/pull/6849

**Pull request closed without merging**